File: backend/SpringBoot_Service/src/main/java/com/postgresql/SpringBoot_Service/EnrollmentController.java
package com.postgresql.SpringBoot_Service;

import com.postgresql.SpringBoot_Service.model.*;
import com.postgresql.SpringBoot_Service.repo.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.*;

@RestController
@RequestMapping("/api/student")
public class EnrollmentController {

    @Autowired
    private UserRepo userRepo;
    
    @Autowired
    private FacultyStudentRepo facultyStudentRepo;
    
    @Autowired
    private SubjectRepo subjectRepo;
    
    @Autowired
    private EnrollmentRepo enrollmentRepo;

    @Autowired
    private MajorRepo majorRepo;

    @PostMapping("/enroll")
    @Transactional
    public ResponseEntity<?> enrollStudent(@RequestBody Map<String, Object> request) {
        try {
            String email = (String) request.get("email");
            String majorName = (String) request.get("majorName");
            Integer year = (Integer) request.get("year");
            List<Integer> subjectIds = (List<Integer>) request.get("subjects");

            // Validate input
            if (email == null || majorName == null || year == null || subjectIds == null) {
                return ResponseEntity.badRequest().body(Map.of("error", "Missing required fields"));
            }

            // Get the student
            Faculty_users user = userRepo.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

            FacultyStudent student = facultyStudentRepo.findByFacultyUser(user);
            if (student == null) {
                throw new RuntimeException("Student record not found");
            }

            // Find the major first
            Major major = majorRepo.findByName(majorName)
                .orElseThrow(() -> new RuntimeException("Major not found: " + majorName));

            // Update student information
            student.setGodinaStudija(String.valueOf(year));
            student.setSmjerStudija(majorName);
            facultyStudentRepo.save(student);

            // Clear existing enrollments in a separate transaction
            enrollmentRepo.deleteAllByStudent(student);

            // Create new enrollments
            List<StudentEnrollment> newEnrollments = new ArrayList<>();
            
            // Set the academic year (e.g., "2024/2025")
            String academicYear = String.format("%d/%d", LocalDate.now().getYear(), LocalDate.now().getYear() + 1);

            for (Integer subjectId : subjectIds) {
                Subject subject = subjectRepo.findById(subjectId.longValue())
                    .orElseThrow(() -> new RuntimeException("Subject not found: " + subjectId));

                StudentEnrollment enrollment = new StudentEnrollment();
                enrollment.setStudent(student);
                enrollment.setSubject(subject);
                enrollment.setEnrollmentDate(LocalDate.now());
                enrollment.setYear(academicYear);
                enrollment.setMajor(major);

                newEnrollments.add(enrollment);
            }

            // Save all enrollments
            enrollmentRepo.saveAll(newEnrollments);

            Map<String, Object> response = new HashMap<>();
            response.put("message", "Successfully enrolled");
            response.put("major", majorName);
            response.put("year", year);
            response.put("subjectsCount", subjectIds.size());

            return ResponseEntity.ok(response);

        } catch (Exception e) {
            // Log the full stack trace
            e.printStackTrace();
            
            // Return a more detailed error message
            String errorMessage = e.getMessage();
            if (e.getCause() != null) {
                errorMessage += " Caused by: " + e.getCause().getMessage();
            }
            
            return ResponseEntity.badRequest().body(Map.of(
                "error", errorMessage,
                "errorType", e.getClass().getSimpleName()
            ));
        }
    }
}

File: backend/src/controllers/enrollment.controller.ts
import { Request, Response } from "express";
import { EnrollmentService } from "../services/enrollment.service";

export class EnrollmentController {
  /**
   * GET /api/enrollment/majors/with-subjects
   * Get all majors with subjects
   */
  static async getMajorsWithSubjects(_req: Request, res: Response) {
    try {
      const majors = await EnrollmentService.getMajorsWithSubjects();
      return res.json(majors);
    } catch (error) {
      console.error("Get majors error:", error);
      return res.status(500).json({ error: "Failed to fetch majors" });
    }
  }

  /**
   * POST /api/enrollment/enroll
   * Enroll student in year with subjects
   */
  static async enrollStudent(req: Request, res: Response) {
    try {
      const { email, majorName, year, subjects } = req.body;

      if (!email || !majorName || !year || !subjects || subjects.length === 0) {
        return res.status(400).json({ error: "Missing required fields" });
      }

      const result = await EnrollmentService.enrollStudent({
        email,
        majorName,
        year,
        subjects,
      });

      return res.json(result);
    } catch (error: any) {
      console.error("Enrollment error:", error);
      return res
        .status(400)
        .json({ error: error.message || "Enrollment failed" });
    }
  }

  /**
   * GET /api/enrollment/student/:email
   * Get student enrollments
   */
  static async getStudentEnrollments(req: Request, res: Response) {
    try {
      const { email } = req.params;
      const enrollments = await EnrollmentService.getStudentEnrollments(email);

      if (!enrollments) {
        return res.status(404).json({ error: "Student not found" });
      }

      return res.json(enrollments);
    } catch (error) {
      console.error("Get enrollments error:", error);
      return res.status(500).json({ error: "Failed to fetch enrollments" });
    }
  }

  /**
   * DELETE /api/enrollment/:id
   * Delete enrollment
   */
  static async deleteEnrollment(req: Request, res: Response) {
    try {
      const id = parseInt(req.params.id);
      await EnrollmentService.deleteEnrollment(id);
      return res.json({ message: "Enrollment deleted successfully" });
    } catch (error) {
      console.error("Delete enrollment error:", error);
      return res.status(500).json({ error: "Failed to delete enrollment" });
    }
  }
}
